### PR TITLE
Add option to automatically apply hotend offsets to EVENT_GCODE_TOOLCHANGE

### PR DESCRIPTION
**Pull request by ansonl**
### Description

This adds the option for firmware to automatically apply hotend offsets to toolchange gcode with a `EVENT_GCODE_TOOLCHANGE_AUTO_HOTEND_OFFSET` option in `Configuration_adv.h`. This feature fixes the missing link between `HOTEND_OFFSET` and `EVENT_GCODE_TOOLCHANGE` so both hotend offset and toolchange gcode can be changed at runtime and during a prints. Turning on this option makes hotend offsets affect toolchange gcode which simplifies the "offset" calculation and saves users massive time. 

Toolchange gcode to extruders are executed within the offsets of the target hotend. The target hotend is used because the target hotend offset is always known while the previous hotend can be one of multiple hotends and multiple offsets (depending on the number of hotends). 

All hotend offsets can be dynamically changed at runtime with [M218](https://marlinfw.org/docs/gcode/M218.html) or defaults set during compile in `Configuration.h`. 

```
M218 T1 Z-1.5 //Change Z offset of T1 to -1.5

#define HOTEND_OFFSET_X { 0.0, 19.85 } // (mm) relative X-offset for each nozzle
#define HOTEND_OFFSET_Y { 0.0, -0.45 }  // (mm) relative Y-offset for each nozzle
#define HOTEND_OFFSET_Z { 0.0, -1.05 }      // (mm) relative Z-offset for each nozzle
```

Toolchange gcode consists of a sequence of gcode commands of any length. The gcode is set during compile in `Configuration_adv.h` with `EVENT_GCODE_TOOLCHANGE_T1`. After testing toolchange gcode with T0 at zero offsets, users need to manually add hotend offsets in every axis for every individual gcode command in `EVENT_GCODE_TOOLCHANGE_T1`. In the below, the commented `EVENT_GCODE_TOOLCHANGE_T1` can be easily developed and manually tested by the user in a safe zero offset environment (T0). There is currently no way to change toolchange gcode at runtime. The uncommented `EVENT_GCODE_TOOLCHANGE_T1` is the actual gcode that must be hard coded into the firmware and can't be changed by the user while using the printer. When the user calibrates and adjusts their hotend offsets by actively using the printer, they can adjust hotend offset with `M218` but any hotend offset changes are not reflected in `EVENT_GCODE_TOOLCHANGE_T1` so the toolchange gcode will always have the wrong "offset" after using `M218` until the user manually calculates new gcode and has access to a compiler. 

```
#define EVENT_GCODE_TOOLCHANGE_T0 "G90\nG0 X210 F7200\nG0 Y165\nG0 X216\nG0 X216 Y184.5\nG0 X210" // run while executing tool-change command T0
//#define EVENT_GCODE_TOOLCHANGE_T1 "G1 X210.00 F7200\nG1 Y199.00\nG1 X216.00\nG1 X216.00 Y182.00\nG1 X210.00" // run while executing tool-change command T1
#define EVENT_GCODE_TOOLCHANGE_T1 "G0 X229.85 F7200\nG0 Y198.55\nG0 X235.85\nG0 Y185.05\nG0 X229.85" // run while executing tool-change command T1
```

`M218` not useful for users without automatic toolchange gcode adjustment because any hotend offset adjustment requires manual gcode editing and compiling. 

### Requirements

Requires more than 1 extruder with offsets.

### Benefits

See above.

This addresses the issue where users with toolchange gcode must change 2 settings (hotend offset + toolchange gcode) to actually change 1 setting (toolchange with offset hotends).

### Configurations

[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/10773410/Configuration.zip)

### Related Issues

This PR is compatible with #24553

### CODE REVIEW

**Comment by thinkyhead** [comment]
Q: Under what circumstances will a post-toolchange move want to use something other than the nozzle position as its reference? Is this to accommodate moves that are carriage-relative?

Note: Instead of modifying the user's custom G-code before executing it, it would probably be simpler to send a `G92` command with the current position minus the hotend offset ahead of the custom G-code, followed by a `G92` that sets the current position back to its previous value. This simply requires that `NO_WORKSPACE_OFFSETS` be disabled.

**Comment by ansonl** [comment]
@thinkyhead Post toolchange move should use nozzle position as the reference in every case except for moving axes or motors that do not get affected by the change in hotend offset. The toolchange gcode that moves the printhead usually moves the nozzle to specific points relative to the carriage/frame so that the printhead can dock/undock, pickup/drop off, activate lever for tools. I can't think of too many cases where post toolchange move in gcode would not depend on the nozzle position. 

I tried manually adding a `G92` before (subtract offset) and after (add offset) the toolchange and it seems to work with limited testing. First `G92` will need to be added after `do_blocking_move_to_xy` on line 1351 is run in `tool_change()` so the new offset adjusted position can be calculated after the printer logic handles the edge case where the nozzle is already at a position like (0,0) and the nozzle cannot fully shift further left like it normally does for the hotend offset adjustment if the nozzle was somewhere else with space like (100,100). Second `G92` should be added after the toolchange gcode is run. Besides the edge cases where the nozzle doesn't have enough space for the normal hotend offset adjustment move, I'm not sure if there is any way for ` gcode.process_subcommands_now` that runs the toolchange gcode to be interrupted and additional gcode from another source to be run in between?

Is there a way to generate a `G92` command without creating a new command string and send the new positions values directly to the code? I saw the powerloss code do something like `sprintf_P(cmd, PSTR("G92.9E%s"), dtostrf(info.current_position.e, 1, 3, str_1));` but that still creates a new string.

File: Marlin/src/module/tool_change.cpp [hunk 1]
```diff
@@ -1409,14 +1403,61 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
...
+
+      #if ANY(TC_GCODE_USE_GLOBAL_X, TC_GCODE_USE_GLOBAL_Y, TC_GCODE_USE_GLOBAL_Z)
+        // G0/G1/G2/G3/G5 moves are relative to the active tool.
+        // Shift the workspace to make custom moves relative to T0.
+        xyz_pos_t old_position_shift;
+        if (new_tool > 0) {
+          old_position_shift = position_shift;
+          const xyz_pos_t &he = hotend_offset[new_tool];
+          #if ENABLED(TC_GCODE_USE_GLOBAL_X)
+            position_shift.x -= he.x; update_workspace_offset(X_AXIS);
```
Comment (thinkyhead) [inline Marlin/src/module/tool_change.cpp:1415]: I believe this is shifting in the right direction. If not, the `-=` can be changed to `+=` and that should do the trick.

File: Marlin/src/inc/Conditionals_adv.h [hunk 2]
```diff
@@ -1152,3 +1152,10 @@
 #if EITHER(INPUT_SHAPING_X, INPUT_SHAPING_Y)
   #define HAS_SHAPING 1
 #endif
+
+// Toolchange Event G-code
+#if !HAS_MULTI_EXTRUDER || !(defined(EVENT_GCODE_TOOLCHANGE_T0) || defined(EVENT_GCODE_TOOLCHANGE_T1) || defined(EVENT_GCODE_TOOLCHANGE_T2) || defined(EVENT_GCODE_TOOLCHANGE_T3) || defined(EVENT_GCODE_TOOLCHANGE_T4) || defined(EVENT_GCODE_TOOLCHANGE_T5) || defined(EVENT_GCODE_TOOLCHANGE_T6) || defined(EVENT_GCODE_TOOLCHANGE_T7))
```
Comment (ansonl) [inline Marlin/src/inc/Conditionals_adv.h:1157]: @thinkyhead  Should the `defined()` checked for `EVENT_GCODE_TOOLCHANGE_T1` and onwards be `!defined()` instead? If `EVENT_GCODE_TOOLCHANGE_T1` is defined by the user as intended, wouldn't the `TC_GCODE_USE_GLOBAL_X` be disabled?

Comment (ansonl) [reply Marlin/src/inc/Conditionals_adv.h:1157]: Nevermind, I just noticed the extra parentheses `!(` around all the `defined()` calls so it makes sense.

**Comment by ansonl** [comment]
@thinkyhead Tested on the actual printer with 2 extruders and confirmed working.

**Pull request merged**